HarfangLab/harfanglab/download_file_from_endpoint.py

```diff
@@ -0,0 +1,155 @@
+
+
+StrOrUUID4: TypeAlias = str | uuid.UUID
+
+
+class DownloadFileFromEndpointAction(JobExecutor):
+    """
+    Action to download a single file from an HarfangLab endpoint
+    """
+
+    _artefact_info: dict[str, Any] | None = None
+    _artefact_info_fetched: bool = False
+
+    @property
+    def artefact_sha256(self) -> str:
+        if not self._artefact_info_fetched:
+            raise RuntimeError("DownloadFileFromEndpointAction.fetch_artefact_info() not called")
+        if self._artefact_info is None:
+            # Note: This raise is mostly for semantic/mypy purpose and should never
+            # happen in real life (an error will always raise before)
+            raise RuntimeError(
+                "DownloadFileFromEndpointAction.fetch_artefact_info() has been called, "
+                "but '_artifact_info' is still None - Something really strange and wrong happen"
+            )
+        return self._artefact_info["sha256"]
+
+    @tenacity.retry(
+        stop=stop_after_delay(30),  # Aggregation can take some time.
+        wait=wait_exponential(min=0.5, max=5),
+        retry=retry_if_exception_message(match="No artefact info available for job"),
+        reraise=True,
+    )
+    def fetch_artefact_info(self, job_id: StrOrUUID4, agent_id: StrOrUUID4) -> None:
+
+        artefact_info_endpoint: str = urllib.parse.urljoin(
+            base=self.instance_url,
+            url="/api/data/investigation/artefact/Artefact/",
+        )
+
+        params: dict[str, Any] = {
+            "job_id": str(job_id),
+            "agent_id": str(agent_id),
+        }
+
+        response: requests.Response = requests.get(artefact_info_endpoint, params=params, headers=self.auth_headers)
+        response.raise_for_status()
+
+        results_count: int = response.json()["count"]
+
+        if results_count == 0:
+            raise ValueError(
+                f"No artefact info available for job {job_id} - "
+                f"Does the targeted file exist on endpoint {agent_id}?"
+            )
+
+        if results_count > 1:
+            raise ValueError(f"Expected 1 result maximum, got {results_count}")
+
+        # At this point, results_count is granted to be 1,
+        # and response.json()["results"] to have one and only one element.
+        artefact_info = response.json()["results"][0]
+
+        if artefact_info["job_id"] != job_id:
+            raise ValueError(
+                f"Given job id and the one in the fetched artefact info missmatch "
+                f"(expected '{job_id}', got '{artefact_info['job_id']}')"
+            )
+
+        if artefact_info["agent"]["agentid"] != agent_id:
+            raise ValueError(
+                f"Given agent id and the one in the fetched artefact info missmatch "
+                f"(expected '{agent_id}', got '{artefact_info['agent']['agentid']}')"
+            )
+
+        self._artefact_info = artefact_info
```

Comment:
I suggest that the `download_status` property in the result of the job must be verified and raise an error if some wnet wrong.